code/modules/power/chem_combustion.dm

```diff
@@ -0,0 +1,622 @@
+			<hr>
+			<span class='notice'><b>You take readings from the [src]'s engine control system:</b></span><br><br>
+			<span class='notice'>Mix: [src.last_mix]</span><br>
+			<span class='notice'>Inlet Flow: [src.last_inlet * 100]%</span><br>
+			<span class='notice'>Oxygen Purity: [src.last_oxygen * 100]%</span><br>
+			<span class='notice'>Fuel Rating: [src.last_fuel]</span>
+			<hr>
+			"})
+			return
+
+		return
+
+	process(var/mult)
+		if (!src || !src.active)
+			return
+
+		src.last_fuel = src.get_fuel_power(src.fuel_tank.reagents)
+		src.last_oxygen = src.check_available_oxygen()
+		if (!src.last_oxygen || !src.last_fuel || !src.fuel_inlet || !src.atmos_inlet)
+			src.stop_engine()
+			src.visible_message("<span class='alert'>The [src]'s engine fails to run, it has nothing to combust!</span>")
+			src.post_status(null, "command", "error", "data", "COMBUSTION_FAILURE", multicast = 1)
+			return
+
+		if (!src.anchored)
+			src.stop_engine()
+			src.visible_message("<span class='alert'>The [src] makes a horrible racket and shuts down, it has become unanchored!</span>")
+			src.post_status(null, "command", "error", "data", "UNANCHORED", multicast = 1)
+			return
+
+		var/fuel_air_mix = (src.atmos_inlet / src.fuel_inlet) // difference between current mix and optimal mix.
+		if (fuel_air_mix < 8 || fuel_air_mix > 20) // too much or too little air or fuel
+			src.stop_engine()
+			src.visible_message("<span class='alert'>The [src] sputters for a moment and then stops, it failed to combust the reagents.</span>")
+			src.post_status(null, "command", "error", "data", "INVALID_MIX", multicast = 1)
+			return
+
+		var/obj/cable/C = src.get_output_cable()
+		if (!C)
+			src.stop_engine()
+			src.visible_message("<span class='alert'>Electricity begins to arc off the [src] causing it to shutdown, it has nothing to output to!</span>")
+			src.post_status(null, "command", "error", "data", "NO_POWER_OUTLET", multicast = 1)
+			elecflash(src.loc, 0, power = 3, exclude_center = 0)
+			return
+
+		src.last_mix = OPTIMAL_MIX / fuel_air_mix
+		src.last_inlet = (src.fuel_inlet + src.atmos_inlet) / 2
+		var/oxygen_multiplier = clamp(src.last_oxygen * 5, 0, 3)
+
+		src.last_output = (src.last_fuel * (src.last_mix * 2) * src.last_inlet * oxygen_multiplier * src.output_multiplier * mult) KILO WATTS
+		var/datum/powernet/P = C.get_powernet()
+		P.newavail += src.last_output
+
+		var/turf/simulated/T = get_turf(src.loc)
+		if (istype(T))
+			var/datum/gas_mixture/payload = new /datum/gas_mixture
+			payload.carbon_dioxide = CARBON_OUTPUT_RATE * (src.last_mix * 2) * src.last_inlet * src.output_multiplier * mult
+			payload.temperature = 323.15 // bit hotter since its exhaust
+			T.assume_air(payload)
+
+		if (src.check_tank_oxygen(src.inlet_tank))
+			src.inlet_tank.remove_air(ATMOS_DRAIN_RATE * (src.last_mix * 2) * src.last_inlet * src.output_multiplier * mult)
+		else if (istype(T))
+			T.remove_air(ATMOS_DRAIN_RATE * (src.last_mix * 2) * src.last_inlet * mult)
```

Comment:
Due to order of operations doesn't this remove some of the CO2 that was just added?

Preferably it would get air from room or tank, combust and thus remove some oxygen and add CO2, then add gas back to room.